torchtune/utils/_distributed.py

```diff
@@ -242,6 +250,83 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
     return lora_wrap_fsdp
 
 
+def load_from_full_model_state_dict(
+    model: FSDPModule,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+    is_rank_zero: bool,
+):
+    meta_sharded_sd = model.state_dict()
+    sharded_sd = {}
+    for param_name, full_tensor in full_sd.items():
+        sharded_meta_param = meta_sharded_sd.get(param_name)
+        # `.to(dtype)` ensures same dtype when `assign=True`
+        sharded_tensor = distribute_tensor(
+            full_tensor.to(sharded_meta_param.dtype),
+            sharded_meta_param.device_mesh,
+            sharded_meta_param.placements,
+        )
+        sharded_sd[param_name] = nn.Parameter(sharded_tensor)
+    # choose `assign=True` since we cannot call `copy_` on meta tensor
+    model.load_state_dict(sharded_sd, strict=False, assign=True)
+
+
+def get_full_model_state_dict(
+    model: FSDPModule,
+    is_rank_zero: bool,
+) -> Dict[str, Any]:
+    sharded_sd = model.state_dict()
+    cpu_state_dict = {}
+    for param_name, sharded_param in sharded_sd.items():
+        full_param = sharded_param.full_tensor()
+        if is_rank_zero:
+            cpu_state_dict[param_name] = full_param.cpu()
+        else:
+            del full_param
+    return cpu_state_dict
+
+
+def load_from_full_optimizer_state_dict(
+    opt: Optimizer,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+) -> Dict[str, Any]:
+    PARAMS = "params"  # noqa: N806
+    _init_optim_state(opt)
+    param_groups = opt.state_dict()["param_groups"]
+    state = opt.state_dict()["state"]
+
+    full_param_groups = full_sd["param_groups"]
+    full_state = full_sd["state"]
+
+    for param_group, full_param_group in zip(param_groups, full_param_groups):
```

Comment:
Do we ever need to worry about proper ordering for param groups (similar q for the zip in L307)?

Reply:
good question! I assume `load_from_full_optimizer_state_dict` is called when resume training. that means as long as the order matches with `get_optimizer_state_dict`, it should be fine.  both functions are under our control

a longer-term solution might be using FQNs to align on order

Reply:
Thanks for the clarification, this sounds like a reasonable assumption to me (we make kinda similar assumptions for other intermediate state saved for checkpoint resume anyways)